imagebuildah/build.go

```diff
@@ -1250,12 +1256,22 @@ func getCreatedBy(node *parser.Node) string {
 // in the Dockerfile till the point of build we are at.
 // Used to verify whether a cache of the intermediate image exists and whether
 // to run the build again.
-func historyMatches(children []*parser.Node, history []v1.History) bool {
+func (b *Executor) historyMatches(children []*parser.Node, history []v1.History) bool {
 	i := len(history) - 1
 	for j := len(children) - 1; j >= 0; j-- {
 		instruction := children[j].Original
 		if children[j].Value == "run" {
 			instruction = instruction[4:]
+			buildArgs := b.getBuildArgs()
+			// If a previous image was built with some build-args but the new build process doesn't have any build-args
+			// specified, so compare the lengths of the old instruction with the current one
+			if buildArgs == "" && len(history[i].CreatedBy) > len(instruction)+11 {
+				return false
+			}
+			// There are build-args, so check if anything with the build-args has changed
+			if buildArgs != "" && !strings.Contains(history[i].CreatedBy, buildArgs) {
+				return false
+			}
```

Comment:
Nice!  Rambling note. I was just looking at this function this weekend in regards to #1391 and found it wasn't working there as I thought it should, although I was approaching it by adding the value from the build args to the applicable arg.  So I changed 'ARG user' to "ARG user=0" if --build-arg user=0 was passed in.  I was tripping over the string contains below with that.  

So end of ramble, will this handle the case where --build-args user=0 on the first run and --build-args=1 on the second run.  Will the code think they should be fresh layers if --layers=true or will will it reuse the cache?

Reply:
yup, it will handle that case. I am adding any build-args used during the process to the history of that layer, as docker does as well. So this line compares the build-args if build-args are set during the process.
I also have a check where the image was built with a build-arg, but in the second run it wasn't. And when that happens, we want a fresh layer instead of cache.

Reply:
Thanks @umohnani8 !  The code LGTM on a quick read, but I want to test/play with it.  Looks like Red Hat CI has gone off the rails.  Might be worth a new push.

Reply:
thanks! and re pushed, hopefully the CI will not die again